### PR TITLE
Fix @objc checking for optional value types

### PR DESCRIPTION
**Pull request by jrose-apple**
We previously treated `NSFoo?` as ObjC-compatible simply because `NSFoo` was an imported type or was marked `@objc`.

---

Before merging this pull request to apple/swift repository:
- [ ] Test pull request on Swift continuous integration.

<details>
  <summary>

Triggering Swift CI</summary>

The swift-ci is triggered by writing a comment on this PR addressed to the GitHub user @swift-ci. Different tests will run depending on the specific comment that you use. The currently available comments are:

**Smoke Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please smoke test |
| OS X platform | @swift-ci Please smoke test OS X platform |
| Linux platform | @swift-ci Please smoke test Linux platform |

 **Validation Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please test |
| OS X platform | @swift-ci Please test OS X platform |
| Linux platform | @swift-ci Please test Linux platform |

Note: Only members of the Apple organization can trigger swift-ci.
</details>

### CODE REVIEW

**Comment by jrose-apple** [comment]
@swift-ci Please test

**Comment by jrose-apple** [comment]
@DougGregor, does the second commit look correct? @jckarter, does the third commit look correct?

File: lib/SILGen/SILGenExpr.cpp [hunk 1]
```diff
@@ -2296,25 +2296,14 @@ RValue RValueEmitter::visitRebindSelfInConstructorExpr(
...
-  
-  auto selectorTy = M.Types.getSelectorType();
-  if (selectorTy && ty->isEqual(selectorTy))
-    return true;
-  
-  return false;
+  // Other types like UnsafePointer can also be nullable.
+  ty = OptionalType::get(ty);
+  return ty->isTriviallyRepresentableIn(ForeignLanguage::C,
+                                        M.getAssociatedContext());
```
Comment (jckarter) [inline lib/SILGen/SILGenExpr.cpp]: That's clever. LGTM

File: test/attr/attr_objc.swift [hunk 2]
```diff
@@ -1121,6 +1122,7 @@ class infer_instanceVar1 {
 // CHECK-LABEL: @objc var var_Optional12: OpaquePointer?
 // CHECK-LABEL: @objc var var_Optional13: UnsafeMutablePointer<Int>?
 // CHECK-LABEL: @objc var var_Optional14: UnsafeMutablePointer<Class_ObjC1>?
+// CHECK-LABEL: @objc var var_Optional15: NSRange?
```
Comment (DougGregor) [inline test/attr/attr_objc.swift]: How does an optional NSRange bridge? It seems like this line should trigger an error...

Comment (jrose-apple) [inline test/attr/attr_objc.swift]: As an optional NSValue. I actually went round and round trying to figure out what was wrong with my logic until I remembered that those are bridged.

**Comment by jrose-apple** [comment]
I don't *think* that's me but I'll run the test again, and talk to @milseman if it fails again.

@swift-ci Please test OS X platform

File: test/attr/attr_objc.swift (hunk 2 above)
Comment (jrose-apple) [inline test/attr/attr_objc.swift]: …and now I take it back, because we don't actually bridge them _correctly._ Will fix.

Comment (jrose-apple) [inline test/attr/attr_objc.swift]: Or rather, will forbid, and someone later can decide whether it's supposed to work.

Comment (jrose-apple) [inline test/attr/attr_objc.swift]: Forbidden in 670f7b5.

**Pull request merged**